src/mod_muc_light_utils.erl

```diff
@@ -246,90 +240,113 @@ value2b(Val, float) -> float_to_binary(Val).
-                 false ->
-                     %% Owner is an old owner
-                     lists:keydelete(Owner, 1, AffUsersChanged0);
-                 true ->
-                     %% Owner is a newcomer, should be promoted
-                     lists:keyreplace(Owner, 1, AffUsersChanged0, {Owner, owner})
-             end};
-        _ ->
-            {NewAffUsers0, AffUsersChanged0}
-    end,
-    {ok, NewAffUsers, AffUsersChanged, JoiningUsers, LeavingUsers};
-fix_edge_aff_users_case({error, bad_request}) ->
-    {error, bad_request}.
+maybe_select_new_owner({ok, AU, AUC, JoiningUsers, LeavingUsers} = AffRes) ->
+    {AffUsers, AffUsersChanged} =
+        case {lists:keyfind(owner, 2, AU), find_new_owner(AffRes)} of
+            {false, {NewOwner, PromotionType}} -> %select new owner
+                NewAU = lists:keyreplace(NewOwner, 1, AU, {NewOwner, owner}),
+                NewAUC = case PromotionType of
+                             promote_old_member ->
+                                 [{NewOwner, owner} | AUC];
+                             promote_joined_member ->
+                                 lists:keyreplace(NewOwner, 1, AUC, {NewOwner, owner});
+                             promote_demoted_owner ->
+                                 lists:keydelete(NewOwner, 1, AUC)
+                         end,
+                {NewAU, NewAUC};
+
+            _ ->
+                {AU, AUC}
+        end,
+    {ok, AffUsers, AffUsersChanged, JoiningUsers, LeavingUsers};
+maybe_select_new_owner(Error) ->
+    Error.
+
+-spec find_new_owner(ChangeResult :: change_aff_success()) ->
+    {jid:simple_bare_jid(), promotion_type()} | false.
+find_new_owner({ok, AU, AUC, JoiningUsers, _LeavingUsers}) ->
+    AllMembers = [U || {U, member} <- (AU)],
+    NewMembers = [U || {U, member} <- (AUC)],
+    OldMembers = AllMembers -- NewMembers,
+    DemotedOwners = NewMembers -- JoiningUsers,
+    %% try to select the new owner from:
+    %%   1) old unchanged room members
+    %%   2) new just joined room members
+    %%   3) demoted room owners
+    case {OldMembers, JoiningUsers, DemotedOwners} of
+        {[U | _], _, _} -> {U, promote_old_member};
+        {_, [U | _], _} -> {U, promote_joined_member};
+        {_, _, [U | _]} -> {U, promote_demoted_owner};
+        _ -> false
+    end.
 
--spec apply_aff_users_change(AffUsers :: aff_users(),
-                             AffUsersChanges :: aff_users(),
-                             NewOwner :: new_owner_flag()) ->
+-spec maybe_demote_old_owner(ChangeResult :: change_aff_success() | {error, bad_request}) ->
     change_aff_success() | {error, bad_request}.
-apply_aff_users_change(AU, AUC, NO) ->
-    apply_aff_users_change(AU, [], AUC, [], NO, [], []).
+maybe_demote_old_owner({ok, AU, AUC, JoiningUsers, LeavingUsers}) ->
+    Owners = [U || {U, owner} <- AU],
+    PromotedOwners = [U || {U, owner} <- AUC],
+    OldOwners = Owners -- PromotedOwners,
+    if
```

Comment:
Among the whole mongooseim codebase we don't use if's. That's also a recommendation from elvis.